### PR TITLE
phpDocumentor: init at 3.3.1

### PR DESCRIPTION
**Pull request by dweipert-3138720606**
###### Description of changes

Adds https://phpdoc.org/ at version 3.3.1.
I added it to tools/misc.

I thought if https://github.com/NixOS/nixpkgs/pull/217207 `phpunit` should go to `development/tools/misc`, maybe `phpDocumentor` should just go into `tools/misc.` Don't know really. Feels like there should be a "php tools" folder somewhere, since `phpDocumentor` obviously only works for PHP.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/misc/phpDocumentor/default.nix [hunk 1]
```diff
@@ -0,0 +1,32 @@
+{ stdenv, fetchurl, makeWrapper, lib, php }:
+let
+  pname = "phpDocumentor";
```
Comment (drupol) [inline pkgs/tools/misc/phpDocumentor/default.nix:3]: There's no need to have those in a `let`...`in` statement. You can directly add them in `stdenv.mkDerivation`.
If you need to use `${version`} them multiple times, add the `rec` keyword just after `stdenv.mkDerivation`.

Take inspiration from: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/php-packages/grumphp/default.nix

**Comment by drupol** [comment]
Hi !

Thanks for your contribution :)

I added a minor comment, after that it should be ok for me.

File: pkgs/tools/misc/phpDocumentor/default.nix [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    url = "https://github.com/phpDocumentor/phpDocumentor/releases/download/v${version}/phpDocumentor.phar";
+    sha256 = "SpPSeP1FgfF3YJAxNNhfzePUDZP3OcjGSPPtAsnD57s=";
+  };
+
+  dontUnpack = true;
+
+  nativeBuildInputs = [ makeWrapper ];
+
+  installPhase = ''
+    mkdir -p $out/bin
```
Comment (drupol) [inline pkgs/tools/misc/phpDocumentor/default.nix:19]: The `preInstall`  and `postInstall` hooks are missing.

See: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/php-packages/grumphp/default.nix

**Review by drupol — CHANGES_REQUESTED** [review]

**Comment by DRogueRonin** [comment]
I `amend`ed and pushed the changes!

**Review by drupol — APPROVED** [review]
LGTM

File: pkgs/tools/misc/phpDocumentor/default.nix (hunk 1 above)
Comment (drupol) [reply pkgs/tools/misc/phpDocumentor/default.nix:4]: Now it's perfectly fine, the latest changes are ok.

File: pkgs/tools/misc/phpDocumentor/default.nix (hunk 2 above)
Comment (drupol) [reply pkgs/tools/misc/phpDocumentor/default.nix:18]: Fixed.

**Review by drupol — APPROVED** [review]

File: pkgs/tools/misc/phpDocumentor/default.nix [hunk 3]
```diff
@@ -0,0 +1,32 @@
+{ stdenv, fetchurl, makeWrapper, lib, php }:
+
+stdenv.mkDerivation rec {
```
Comment (drupol) [inline pkgs/tools/misc/phpDocumentor/default.nix:3]: I think we should use the `finalAttrs` pattern here.

Let me share a couple of useful links regarding this:

1. History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
2. Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
3. Recent example of implementation: https://github.com/NixOS/nixpkgs/compare/17f96f7b978e61576cfe16136eb418f74fab9952..9e6ea843e473d34d4f379b8b0d8ef0425a06defe

Feel free to come back to me if you need further help.

Cheers.

**Review by drupol — CHANGES_REQUESTED** [review]

**Comment by drupol** [comment]
Just FYI, I added `phpDocumentor` in this PR () at this specific commit: https://github.com/NixOS/nixpkgs/pull/225401/commits/a2f2ad40bdb48a0ec54402376e883f911ed10e0a

An upstream patch/fix had to be created: https://github.com/phpDocumentor/phpDocumentor/pull/3518

While I'm not against merging this PR in the meantime, feel free to let me know if you wish to continue the work on it.

**Comment by DRogueRonin** [comment]
No problem for me at all to wait for your PR and to close this one. I'll just keep the notion about the `finalAttrs` in mind for the next time I contribute a package. :)

**Comment by drupol** [comment]
Great! Thank you :)

**Pull request closed without merging**